formal-models/dbft/dbft.tla

```diff
@@ -0,0 +1,368 @@
+                                  /\ msg.view = rmState[r].view
+                 }) >= M
+  /\ PrepareRequestSentOrReceived(r)
+  /\ rmState' = [rmState EXCEPT ![r].type = "commitSent"]
+  /\ msgs' = msgs \cup {[type |-> "Commit", rm |-> r, view |-> rmState[r].view]}
+  /\ UNCHANGED <<>>
+
+\* RMAcceptBlock describes node r collecting enough Commit messages and accepting
+\* the block.
+RMAcceptBlock(r) ==
+  /\ rmState[r].type /= "bad"
+  /\ rmState[r].type /= "dead"
+  /\ PrepareRequestSentOrReceived(r)
+  /\ Cardinality({msg \in msgs : msg.type = "Commit" /\ msg.view = rmState[r].view}) >= M
+  /\ rmState' = [rmState EXCEPT ![r].type = "blockAccepted"]
+  /\ UNCHANGED <<msgs>>
+
+\* RMSendChangeView describes node r sending ChangeView message on timeout.
+RMSendChangeView(r) ==
+  /\ \/ rmState[r].type = "initialized"
+     \/ rmState[r].type = "prepareSent"
+  /\ LET cv == [type |-> "ChangeView", rm |-> r, view |-> rmState[r].view]
+     IN /\ cv \notin msgs
+        /\ rmState' = [rmState EXCEPT ![r].type = "cv"]
+        /\ msgs' = msgs \cup {[type |-> "ChangeView", rm |-> r, view |-> rmState[r].view]}
+
+\* RMReceiveChangeView describes node r receiving enough ChangeView messages for
+\* view changing.
+RMReceiveChangeView(r) ==
+  /\ rmState[r].type /= "bad"
+  /\ rmState[r].type /= "dead"
+  /\ rmState[r].type /= "blockAccepted"
+  /\ rmState[r].type /= "commitSent"
+  /\ Cardinality({
+                  rm \in RM : Cardinality({
+                                            msg \in msgs : /\ msg.type = "ChangeView"
+                                                           /\ msg.rm = rm
+                                                           /\ GetNewView(msg.view) >= GetNewView(rmState[r].view)
+                                         }) /= 0
+                 }) >= M
+  /\ rmState' = [rmState EXCEPT ![r].type = "initialized", ![r].view = GetNewView(rmState[r].view)]
+  /\ UNCHANGED <<msgs>>
+
+\* RMBeBad describes the faulty node r that will send any kind of consensus message starting
+\* from the step it's gone wild. This step is enabled only when RMFault is non-empty set.
+RMBeBad(r) ==
+  /\ r \in RMFault
+  /\ Cardinality({rm \in RM : rmState[rm].type = "bad"}) < F
+  /\ rmState' = [rmState EXCEPT ![r].type = "bad"]
+  /\ UNCHANGED <<msgs>>
+
+\* RMFaultySendCV describes sending CV message by the faulty node r.
+RMFaultySendCV(r) ==
+  /\ rmState[r].type = "bad"
+  /\ LET cv == [type |-> "ChangeView", rm |-> r, view |-> rmState[r].view]
+     IN /\ cv \notin msgs
+        /\ msgs' = msgs \cup cv
+        /\ UNCHANGED <<rmState>>
+
+\* RMFaultyDoCV describes view changing by the faulty node r.
+RMFaultyDoCV(r) ==
+  /\ rmState[r].type = "bad"
+  \* TODO: why do we have to check CV not in msgs?
+  /\ [type |-> "ChangeView", rm |-> r, view |-> rmState[r].view] \notin msgs
+  /\ rmState' = [rmState EXCEPT ![r].view = GetNewView(rmState[r].view)]
+  /\ UNCHANGED <<msgs>>
```

Comment:
@roman-khimov, could you, please, explain, what was the idea of adding `  /\ [type |-> "ChangeView", rm |-> r, view |-> rmState[r].view] \notin msgs` line to this action? IMO, it is not needed here, faulty node can do CV whenever it wants.